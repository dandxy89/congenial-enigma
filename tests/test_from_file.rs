use std::path::PathBuf;

use lp_parser_rs::{
    model::{lp_error::LPParserError, lp_problem::LPProblem},
    parse::{parse_file, parse_lp_file},
};

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr, $o_sum:expr, $c_sum:expr) => {
        #[test]
        fn $test_name() {
            let result = read_file_from_resources($file).unwrap();

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(result, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });

            let summation = &result.objectives.iter().map(|o| o.coefficients.iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($o_sum, *summation, abs <= 1e-2);

            let summation = &result.constraints.iter().map(|(_, c)| c.coefficients().iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($c_sum, *summation, abs <= 1e-2);
        }
    };
}

generate_test!(afiro, "afiro.lp", -4.100, 25.369);
generate_test!(afiro_ext, "afiro_ext.lp", -5.239, 25.369);
generate_test!(boeing1, "boeing1.lp", 1187.985, 194612.346);
generate_test!(boeing2, "boeing2.lp", 78.488, 20863.836);
generate_test!(fit1d, "fit1d.lp", 82457.0, -146871.180);
generate_test!(kb2, "kb2.lp", 11.675, 10143.724);
generate_test!(pulp, "pulp.lp", 1., 73.44);
generate_test!(pulp2, "pulp2.lp", -6.0, 147.);
generate_test!(sc50a, "sc50a.lp", -1., 30.3);
generate_test!(model2, "model2.lp", 0., 6.);
generate_test!(limbo, "limbo.lp", 2., 0.);
generate_test!(obj3_2cons, "3obj_2cons.lp", 1., -186.);
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp", -7.5, -186.);
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp", -7.5, -186.);
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp", -1., 16.5);
generate_test!(semi_continuous, "semi_continuous.lp", 2., 0.);
generate_test!(sos, "sos.lp", 0., 17.5);
generate_test!(test, "test.lp", 2., 2.9899);
generate_test!(test2, "test2.lp", -6., 147.);
generate_test!(empty_bounds, "empty_bounds.lp", 11., 2.);
generate_test!(blank_lines, "blank_lines.lp", 11., 2.);
generate_test!(optional_labels, "optional_labels.lp", 11., 1.);
generate_test!(infile_comments, "infile_comments.lp", 43., 7.);
generate_test!(infile_comments2, "infile_comments2.lp", 43., 0.);
generate_test!(missing_signs, "missing_signs.lp", 43., -3.);
generate_test!(fit2d, "fit2d.lp", 349048.9, -296677.389);
generate_test!(scientific_notation, "scientific_notation.lp", 238.0324, -43.87680);

#[test]
fn invalid() {
    let result = read_file_from_resources("invalid.lp");
    assert!(result.is_err());
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
