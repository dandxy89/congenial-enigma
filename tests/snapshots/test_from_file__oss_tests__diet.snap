---
source: tests/test_from_file.rs
expression: result
---
constraints:
  "c1:":
    Standard:
      name: "c1:"
      coefficients:
        - coefficient: 277.4
          var_name: Roasted_Chicken
        - coefficient: 358.2
          var_name: Spaghetti_W__Sauce
        - coefficient: 25.8
          var_name: "Tomato,Red,Ripe,Raw"
        - coefficient: 81.4
          var_name: "Apple,Raw,W_Skin"
        - coefficient: 15.1
          var_name: Grapes
        - coefficient: 78.1
          var_name: Chocolate_Chip_Cookies
        - coefficient: 121.2
          var_name: Lowfat_Milk
        - coefficient: 115.1
          var_name: Raisin_Brn
        - coefficient: 242.1
          var_name: Hotdog
        - coefficient: -1
          var_name: Rgc1
      sense: "="
      rhs: 2500
  "c2:":
    Standard:
      name: "c2:"
      coefficients:
        - coefficient: 21.9
          var_name: Roasted_Chicken
        - coefficient: 80.2
          var_name: Spaghetti_W__Sauce
        - coefficient: 6.2
          var_name: "Tomato,Red,Ripe,Raw"
        - coefficient: 9.7
          var_name: "Apple,Raw,W_Skin"
        - coefficient: 3.4
          var_name: Grapes
        - coefficient: 6.2
          var_name: Chocolate_Chip_Cookies
        - coefficient: 296.7
          var_name: Lowfat_Milk
        - coefficient: 12.9
          var_name: Raisin_Brn
        - coefficient: 23.5
          var_name: Hotdog
        - coefficient: -1
          var_name: Rgc2
      sense: "="
      rhs: 1600
  "c3:":
    Standard:
      name: "c3:"
      coefficients:
        - coefficient: 1.8
          var_name: Roasted_Chicken
        - coefficient: 2.3
          var_name: Spaghetti_W__Sauce
        - coefficient: 0.6
          var_name: "Tomato,Red,Ripe,Raw"
        - coefficient: 0.2
          var_name: "Apple,Raw,W_Skin"
        - coefficient: 0.1
          var_name: Grapes
        - coefficient: 0.4
          var_name: Chocolate_Chip_Cookies
        - coefficient: 0.1
          var_name: Lowfat_Milk
        - coefficient: 16.8
          var_name: Raisin_Brn
        - coefficient: 2.3
          var_name: Hotdog
        - coefficient: -1
          var_name: Rgc3
      sense: "="
      rhs: 30
  "c4:":
    Standard:
      name: "c4:"
      coefficients:
        - coefficient: 77.4
          var_name: Roasted_Chicken
        - coefficient: 3055.2
          var_name: Spaghetti_W__Sauce
        - coefficient: 766.3
          var_name: "Tomato,Red,Ripe,Raw"
        - coefficient: 73.1
          var_name: "Apple,Raw,W_Skin"
        - coefficient: 24
          var_name: Grapes
        - coefficient: 101.8
          var_name: Chocolate_Chip_Cookies
        - coefficient: 500.2
          var_name: Lowfat_Milk
        - coefficient: 1250.2
          var_name: Raisin_Brn
        - coefficient: -1
          var_name: Rgc4
      sense: "="
      rhs: 50000
objectives:
  - coefficients:
      - coefficient: 0.84
        var_name: Roasted_Chicken
      - coefficient: 0.78
        var_name: Spaghetti_W__Sauce
      - coefficient: 0.27
        var_name: "Tomato,Red,Ripe,Raw"
      - coefficient: 0.24
        var_name: "Apple,Raw,W_Skin"
      - coefficient: 0.32
        var_name: Grapes
      - coefficient: 0.03
        var_name: Chocolate_Chip_Cookies
      - coefficient: 0.23
        var_name: Lowfat_Milk
      - coefficient: 0.34
        var_name: Raisin_Brn
      - coefficient: 0.31
        var_name: Hotdog
    name: obj
problem_name: ~
problem_sense: Minimize
variables:
  "Apple,Raw,W_Skin": General
  Chocolate_Chip_Cookies: General
  Grapes: General
  Hotdog: General
  Lowfat_Milk: General
  Raisin_Brn: General
  Rgc1: General
  Rgc2: General
  Rgc3: General
  Rgc4: General
  Roasted_Chicken: General
  Spaghetti_W__Sauce: General
  "Tomato,Red,Ripe,Raw": General
